Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Basic.lean

```diff
@@ -0,0 +1,427 @@
+lemma ΨSq_zero : W.ΨSq 0 = 0 := by
+  erw [ΨSq_ofNat, preΨ'_zero, zero_pow two_ne_zero, zero_mul]
+
+@[simp]
+lemma ΨSq_one : W.ΨSq 1 = 1 := by
+  erw [ΨSq_ofNat, preΨ'_one, one_pow, mul_one]
+
+@[simp]
+lemma ΨSq_two : W.ΨSq 2 = W.Ψ₂Sq := by
+  erw [ΨSq_ofNat, preΨ'_two, one_pow, one_mul, if_pos even_two]
+
+@[simp]
+lemma ΨSq_three : W.ΨSq 3 = W.Ψ₃ ^ 2 := by
+  erw [ΨSq_ofNat, preΨ'_three, mul_one]
+
+@[simp]
+lemma ΨSq_four : W.ΨSq 4 = W.preΨ₄ ^ 2 * W.Ψ₂Sq := by
+  erw [ΨSq_ofNat, preΨ'_four, if_pos <| by decide]
+
+lemma ΨSq_odd (m : ℕ) : W.ΨSq (2 * (m + 2) + 1) =
+    (W.preΨ' (m + 4) * W.preΨ' (m + 2) ^ 3 * (if Even m then W.Ψ₂Sq ^ 2 else 1) -
+      W.preΨ' (m + 1) * W.preΨ' (m + 3) ^ 3 * (if Even m then 1 else W.Ψ₂Sq ^ 2)) ^ 2 := by
+  erw [ΨSq_ofNat, preΨ'_odd, if_neg (m + 2).not_even_two_mul_add_one, mul_one]
+
+lemma ΨSq_even (m : ℕ) : W.ΨSq (2 * (m + 3)) =
+    (W.preΨ' (m + 2) ^ 2 * W.preΨ' (m + 3) * W.preΨ' (m + 5) -
+      W.preΨ' (m + 1) * W.preΨ' (m + 3) * W.preΨ' (m + 4) ^ 2) ^ 2 * W.Ψ₂Sq := by
+  erw [ΨSq_ofNat, preΨ'_even, if_pos <| even_two_mul _]
+
+@[simp]
+lemma ΨSq_neg (n : ℤ) : W.ΨSq (-n) = W.ΨSq n := by
+  simp only [WeierstrassCurve.ΨSq, preΨ_neg, neg_sq, even_neg]
+
+/-! ### The bivariate polynomials $\Psi_n$ -/
+
+/-- The bivariate polynomials $\Psi_n$ congruent to the $n$-division polynomials $\psi_n$. -/
+protected noncomputable def Ψ (n : ℤ) : R[X][Y] :=
+  C (W.preΨ n) * if Even n then W.ψ₂ else 1
+
+@[simp]
+lemma Ψ_ofNat (n : ℕ) : W.Ψ n = C (W.preΨ' n) * if Even n then W.ψ₂ else 1 := by
+  simp only [WeierstrassCurve.Ψ, preΨ_ofNat, Int.even_coe_nat]
+
+@[simp]
+lemma Ψ_zero : W.Ψ 0 = 0 := by
+  erw [Ψ_ofNat, preΨ'_zero, C_0, zero_mul]
+
+@[simp]
+lemma Ψ_one : W.Ψ 1 = 1 := by
+  erw [Ψ_ofNat, preΨ'_one, C_1, mul_one]
+
+@[simp]
+lemma Ψ_two : W.Ψ 2 = W.ψ₂ := by
+  erw [Ψ_ofNat, preΨ'_two, one_mul, if_pos even_two]
+
+@[simp]
+lemma Ψ_three : W.Ψ 3 = C W.Ψ₃ := by
+  erw [Ψ_ofNat, preΨ'_three, mul_one]
+
+@[simp]
+lemma Ψ_four : W.Ψ 4 = C W.preΨ₄ * W.ψ₂ := by
+  erw [Ψ_ofNat, preΨ'_four, if_pos <| by decide]
+
+lemma Ψ_odd (m : ℕ) : W.Ψ (2 * (m + 2) + 1) =
+    W.Ψ (m + 4) * W.Ψ (m + 2) ^ 3 - W.Ψ (m + 1) * W.Ψ (m + 3) ^ 3 +
+      W.toAffine.polynomial * (16 * W.toAffine.polynomial - 8 * W.ψ₂ ^ 2) * C
+        (if Even m then W.preΨ' (m + 4) * W.preΨ' (m + 2) ^ 3
+          else -W.preΨ' (m + 1) * W.preΨ' (m + 3) ^ 3) := by
+  repeat erw [Ψ_ofNat]
+  simp_rw [preΨ'_odd, if_neg (m + 2).not_even_two_mul_add_one, Nat.even_add_one, ite_not]
+  split_ifs <;> C_simp <;> rw [C_Ψ₂Sq_eq] <;> ring1
+
+lemma Ψ_even (m : ℕ) : W.Ψ (2 * (m + 3)) * W.ψ₂ =
+    W.Ψ (m + 2) ^ 2 * W.Ψ (m + 3) * W.Ψ (m + 5) - W.Ψ (m + 1) * W.Ψ (m + 3) * W.Ψ (m + 4) ^ 2 := by
+  repeat erw [Ψ_ofNat]
+  simp_rw [preΨ'_even, if_pos <| even_two_mul _, Nat.even_add_one, ite_not]
+  split_ifs <;> C_simp <;> ring1
+
+@[simp]
+lemma Ψ_neg (n : ℤ) : W.Ψ (-n) = -W.Ψ n := by
+  simp only [WeierstrassCurve.Ψ, preΨ_neg, C_neg, neg_mul (α := R[X][Y]), even_neg]
```

Comment:
I've observed a similar issue e.g. [here](https://github.com/leanprover-community/mathlib4/blob/bd3b187a0c6d5cf7fadef7ff51f86faa16a639c2/Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial.lean#L572). I think maybe some previous step only requires a semiring, and afterwards Lean forgets it's actually a ring, maybe due to some unification/instance caching issues.